modules/dag/src/main/scala/org/tesselation/dag/l1/DAGBlockValidator.scala

```diff
@@ -0,0 +1,170 @@
+package org.tesselation.dag.l1
+
+import cats.effect.Async
+import cats.syntax.applicativeError._
+import cats.syntax.either._
+import cats.syntax.flatMap._
+import cats.syntax.functor._
+import cats.syntax.semigroup._
+import cats.syntax.traverse._
+
+import scala.annotation.tailrec
+
+import org.tesselation.dag.l1.storage.{BlockStorage, TransactionStorage}
+import org.tesselation.domain.cluster.storage._
+import org.tesselation.infrastructure.db.schema.StoredAddress
+import org.tesselation.kryo.KryoSerializer
+import org.tesselation.schema.balance.Balance
+import org.tesselation.schema.transaction.{Transaction, TransactionReference}
+import org.tesselation.schema.{BlockValidator, nonNegBigIntSemigroup}
+import org.tesselation.security.signature.Signed
+import org.tesselation.security.{Hashed, SecurityProvider}
+
+import eu.timepit.refined.auto.autoInfer
+import eu.timepit.refined.numeric.NonNegative
+import eu.timepit.refined.refineV
+import io.estatico.newtype.ops.toCoercibleIdOps
+
+class DAGBlockValidator[F[_]: Async: KryoSerializer: SecurityProvider](
+  transactionStorage: TransactionStorage[F],
+  blockStorage: BlockStorage[F],
+  addressStorage: AddressStorage[F],
+  transactionValidator: TransactionValidator[F]
+) extends BlockValidator[F, DAGBlock] {
+
+  override def accept(signedBlock: Signed[DAGBlock]): F[Unit] =
+    for {
+      _ <- validateBlockSignatures(signedBlock).ifM(
+        Async[F].unit,
+        Async[F].raiseError[Unit](new Throwable("Invalid block signatures!"))
+      )
+      _ <- validateBlock(signedBlock.value).ifM(
+        Async[F].unit,
+        Async[F].raiseError[Unit](new Throwable("Invalid block data!"))
+      )
+      hashedBlock <- signedBlock.hashWithSignatureCheck.flatMap(_.liftTo[F])
```

Comment:
lift swallows errors. probably it should be attemptT, rethrowT, or sth like that

Reply:
Ok I will check that and refactor.

Reply:
Checked it and it seems the error is not being swallowed. `liftTo` for `Either` leads to this piece of code and it looks like it rethrows by using `ApplicativeError` instance
```
  def fromEither[A](x: E Either A): F[A] =
    x match {
      case Right(a) => pure(a)
      case Left(e)  => raiseError(e)
    }
```